java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -412,7 +412,7 @@ public void accept(SymbolicValue symbolicValue, PMap<Class<? extends Constraint>
   public ProgramState resetFieldValues(ConstraintManager constraintManager) {
     final List<VariableTree> variableTrees = new ArrayList<>();
     values.forEach((symbol, symbolicValue) -> {
-      if (isField(symbol)) {
+      if (isField(symbol) && !symbol.isFinal()) {
```

Comment:
:+1: I am bit surprised this didn't lead to more issues being detected in ruling